Query functionality for temporal-web: stack trace and registered queries

jeffschoner opened the pull request:
### Summary

- Adds a built-in `__stack_trace` query for all workflows that returns the stack trace of all suspended fibers for a workflow
- The error message returned when an unknown query is received now matches that of the Go SDK ([source](https://github.com/temporalio/sdk-go/blob/cee305f568f4b7cba5ec1c597d206bd0114289d7/internal/internal_workflow.go#L550))
   - Temporal-web intentionally makes an unknown query `__cadence_web_list`, then parses its error message to extract a list of all registered queries.
   - Matching the error message has the effect of enableing a UI in temporal-web for executing queries on the 'Query' tab of a workflow.

### Screenshot

If there are multiple fibers suspended, it displays more than one stack trace. Here, we see a workflow that is sleeping in an activity's done callback, while waiting for an activity to synchronously complete.

<img width="637" alt="image" src="https://user-images.githubusercontent.com/15093535/172080616-ea039b1f-f6eb-4ebe-87ab-a46f71e37058.png">

### Testing

There are several new or modified specs:
```
# Modified spec with some additional tests for the stack trace query end-to-end behavior
bundle exec rspec spec/unit/lib/temporal/workflow/context_spec.rb

# Minorly modified to propagate tracking bit
bundle exec rspec spec/unit/lib/temporal/workflow/executor_spec.rb

# New unit tests for new StackTraceTracker class
bundle exec rspec spec/unit/lib/temporal/workflow/stack_trace_tracker_spec.rb

# Modified existing example test for queries to perform both of these built-in queries and check their outputs
cd examples
bundle rspec examples/spec/integration/query_workflow_spec.rb
```

antstorm reviewed (changes requested):
This is an awesome addition, thank you @jeffschoner! 🙌 

I would like to propose a slightly different approach to handling those queries though (shouldn't be too big of a change, but will simplify things). Basically right now both of the new queries are handled very differently from the user defined queries — each is a special snowflake.

Instead what I think would improve the change is if we processed these using the same exact API as all the other queries. Basically in the `Workflow::Context` on initialization you would register them using:

```ruby
query_registry.register(QueryRegistry::STACK_TRACE_QUERY) do
  stack_trace_tracker.to_s
end

query_registry.register(QueryRegistry::LIST_QUERIES_QUERY) do
  raise "KnownQueryTypes=[#{query_registry.queries.join(', ')}]"
end
```

This should simplify things and make it a bit more straightforward. There's also no harm in tracking the stack trace even if there isn't a stack_trace query (in case it is, you can just not pass one at the `Executor` level) and you don't need to pass it all the way from the `TaskProcessor`

jeffschoner commented on lib/temporal/client.rb:
```diff
@@ -200,6 +201,24 @@ def query_workflow(workflow, query, workflow_id, run_id, *args, namespace: nil,
+    # @param namespace [String, nil] if nil, choose the one declared on the workflow class or the
+    #   global default
+    # @param query_reject_condition [Symbol] check Temporal::Connection::GRPC::QUERY_REJECT_CONDITION
+    def query_workflow_stack_trace(workflow, workflow_id, run_id, namespace: nil, query_reject_condition: nil)
```
I added the one for getting the stack trace since it is straight forward. Calling it returns a stack trace. It could be useful in a variety of circumstances, including just being invoked by tctl to debug.

The list query is not typical. It returns its result by raising an error with a specifically formatted error message. In the Go and Java SDKs, such a query doesn't even exist. They just happen to return this error when the query is unknown. temporal-web then deliberately sends a made up query to scrape the results of the error. This is much more hacky. I could catch the error here and return the list, but it's pushing the bounds of a supported API.

antstorm replied:
@jeffschoner but it's not what's happening here — looking at the implementation this is literally just currying the `query_workflow` with a preset query name. If it's atypical then we need to provide custom handling of the result. Since this is part of a public API we need to think it through before exposing to others (changing it afterwards will be more difficult)

jeffschoner commented:
After taking a step back and thinking about the `__cadence_web_list` pseudo-query more, I decided a better way to support this is to instead update the error message emitted when an unknown query comes in. This dramatically simplifies the code and makes for an overall much more elegant solution. Should have a new revision with this and other changes soon.

jeffschoner commented:
@antstorm Please take another look

I refactored the code to further decouple the task processor and executor from the stack trace tracking. Ultimately, I chose to still toggle the collection of the stack traces on or off based on if it's needed. While this can be left on all the time, it wasn't clear to me how expensive calling `Kernel.caller` is. Fetching the stack trace when it's not going to be needed seems wasteful to me, especially since would be the common case for the overwhelming majority of workflow tasks.

I also got rid of the  query for listing queries in favor of including the same content in the unknown query error message as the Go SDK does. Logically, this makes more sense and cleans up the code quite a bit.

jeffschoner commented:
Oh, I also reorganized my commits so that the 3 atomic changes here are each in their own commit. It's a bit tidier and makes it easier to review commit-by-commit if you prefer.

dwillett commented on lib/temporal/workflow/query_registry.rb:
```diff
@@ -19,7 +19,10 @@ def handle(type, args = nil)
+          # The end of the formatted error message (e.g., "KnownQueryTypes=[query-1, query-2, query-3]")
+          # is used by temporal-web to show a list of queries that can be run on the 'Query' tab of a
+          # workflow. If that part of the error message is changed, that functionality will break.
+          raise Temporal::QueryFailed, "Workflow did not register a handler for '#{type}'. KnownQueryTypes=[#{handlers.keys.join(", ")}]"
```
Nice find - this makes much more sense to me now. I remember seeing the `KnownQueryTypes=...` format included in the Go SDK error message and didn't make the connection. 🤦

jeffschoner replied:
It took me some time to wrap my brain around this as well. Looks like it goes way back to Cadence https://github.com/uber-go/cadence-client/pull/226

jeffschoner commented:
@DeRauk @antstorm Any other thoughts on this one?

antstorm reviewed (changes requested):
Thanks for making the changes, @jeffschoner!

One thing that I'd like to remove from this PR is the new public API (`query_workflow_stack_trace`). I agree that if there's special handling for it required than it does make sense to have it, but it needs to be implemented before the API is exposed since changing it afterwards will be much more difficult. It also lacks unit tests, which we want for a public API despite integration tests.

I suggest dropping it and opening a follow-up PR to discuss and finalize it separately

jeffschoner commented:
> I suggest dropping it and opening a follow-up PR to discuss and finalize it separately

@antstorm This makes sense to me. This is in a separate commit so I should be able to remove it quickly

jeffschoner commented:
@antstorm @DeRauk I've removed the commit, modified another commit where the removed method was still referenced in a spec, and updated the PR description. Unit and example specs pass locally for me against the latest docker-compose image, but I think something is wrong with the Temporal container in CircleCI.

DeRauk commented:
@jeffschoner it might be something wrong with the latest temporal auto-setup image. Can you try pinning it to an older sha? Ex: https://github.com/coinbase/temporal-ruby/pull/187/files

antstorm reviewed (approved):
Looks great! Thank you @jeffschoner !
